File: backend/schemas/horas_cliente.py
"""
Schemas de la vista interna "Horas por cliente" (solo RRHH).

Archivo propio y no `schemas/horas.py`: ahí viven los schemas del REGISTRO (una fila de
`horas_proyecto`) y acá los de una VISTA AGREGADA, que es otra forma — árbol con totales, no
filas. Mezclarlos haría que agregar una métrica a la pantalla toque el schema del registro.
"""
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel

from schemas.horas import HoraResponse, Modalidad


class KPIsHoras(BaseModel):
    """Los cuatro del encabezado. Se calculan sobre el MISMO conjunto que la tabla de abajo."""
    horas_totales: float
    # Clientes REALES: el grupo "Sin cliente" (las cargas del camino viejo) no suma acá.
    clientes_con_carga: int
    empleados_que_cargaron: int
    registros: int


class LineaEmpleado(BaseModel):
    """Una línea del detalle de un cliente: empleado + proyecto + tarea + modalidad, agregados."""
    empleado_id: Optional[UUID] = None
    empleado_nombre: Optional[str] = None
    proyecto_texto: Optional[str] = None
    tarea_texto: Optional[str] = None
    modalidad: Optional[Modalidad] = None
    horas: float
    registros: int


class ClienteConHoras(BaseModel):
    """Un cliente colapsable. `cliente_id` es None en el grupo "Sin cliente"."""
    cliente_id: Optional[UUID] = None
    cliente_nombre: str
    horas: float
    registros: int
    lineas: List[LineaEmpleado]


class HorasPorClienteResponse(BaseModel):
    mes: int
    anio: int
    kpis: KPIsHoras
    clientes: List[ClienteConHoras]


class DetalleEmpleadoResponse(BaseModel):
    """El "ver detalle": las cargas DÍA POR DÍA de un empleado en el período.

    Devuelve `HoraResponse` completo —con el `id`— porque la pantalla necesita el id para poder
    BORRAR. Editar no está: `HorasService` declara los registros inmutables por decisión escrita
    (hay delete y no hay update), y revocarla es una decisión de producto, no una feature.
    """
    items: List[HoraResponse]
    total_horas: float

File: frontend/types/horasCliente.ts
import type { Hora, Modalidad } from "@/types/proyecto"

/** Espejo de backend/schemas/horas_cliente.py. */

export interface KPIsHoras {
  horas_totales: number
  /** Clientes REALES: el grupo "Sin cliente" (cargas del camino viejo) NO suma acá. */
  clientes_con_carga: number
  empleados_que_cargaron: number
  registros: number
}

export interface LineaEmpleado {
  empleado_id: string | null
  empleado_nombre: string | null
  proyecto_texto: string | null
  tarea_texto: string | null
  modalidad: Modalidad | null
  horas: number
  registros: number
}

/** Cuánto puso cada sociedad del grupo contra un cliente. La suma es el total del cliente. */
export interface HorasDeEmpresa {
  empresa_nombre: string
  horas: number
}

export interface ClienteConHoras {
  /** null en el grupo "Sin cliente": son las cargas del camino viejo, que no tienen cliente. */
  cliente_id: string | null
  cliente_nombre: string
  /** Total del CLIENTE, sin recortar por sociedad. `por_empresa` lo reparte. */
  horas: number
  registros: number
  por_empresa: HorasDeEmpresa[]
  lineas: LineaEmpleado[]
}

export interface HorasPorCliente {
  mes: number
  anio: number
  kpis: KPIsHoras
  clientes: ClienteConHoras[]
}

export interface DetalleEmpleado {
  items: Hora[]
  total_horas: number
}
